Warnings about NULL and module operator

I get these warning from clad:
```
test.orig.cpp:18126:76: warning: statement kind 'GNUNullExpr' is not supported
 18126 |     if ((updateConsValsRhoPIndependent(gasModel, blockFamilyMD, stateLeft, NULL, omegaRSqr, invGammaM1, gamma) == EXIT_FAILURE)) {
       |                                                                            ^~~~
/usr/bin/../lib/clang/20/include/__stddef_null.h:21:14: note: expanded from macro 'NULL'
   21 | #define NULL __null
      |              ^~~~~~
test.orig.cpp:18131:138: warning: attempted to differentiate unsupported operator; treated as non-differentiable
 18131 |             {    *(&j) = ((((((block)->mesh.layout.isStructured)))) ? (Funsigned) (((indexNodeLeft) - ((block->mesh.layout).initialPad)) % ((block->mesh.layout).strideLNode) / ((block->mesh.layout).strideJNode)) : (Funsigned) (1)); *(&k) = ((((((block)->mesh.layout.isStructured)))) ? (Funsigned) (((indexNodeLeft) - ((block->mesh.layout).initialPad)) % ((block->mesh.layout).strideJNode)) : (Funsigned) ((indexNodeLeft))); *(&l) = ((((((block)->mesh.layout.isStructured)))) ? (Funsigned) (((indexNodeLeft) - ((block->mesh.layout).initialPad)) / ((block->mesh.layout).strideLNode)) : (Funsigned) (1)); };
       |                                                                                                                                          ^   
test.orig.cpp:18131:353: warning: attempted to differentiate unsupported operator; treated as non-differentiable
 18131 |             {    *(&j) = ((((((block)->mesh.layout.isStructured)))) ? (Funsigned) (((indexNodeLeft) - ((block->mesh.layout).initialPad)) % ((block->mesh.layout).strideLNode) / ((block->mesh.layout).strideJNode)) : (Funsigned) (1)); *(&k) = ((((((block)->mesh.layout.isStructured)))) ? (Funsigned) (((indexNodeLeft) - ((block->mesh.layout).initialPad)) % ((block->mesh.layout).strideJNode)) : (Funsigned) ((indexNodeLeft))); *(&l) = ((((((block)->mesh.layout.isStructured)))) ? (Funsigned) (((indexNodeLeft) - ((block->mesh.layout).initialPad)) / ((block->mesh.layout).strideLNode)) : (Funsigned) (1)); };
       |                                                                                                                                                                                                                                                                                                                                                                 ^
test.orig.cpp:18132:51: warning: statement kind 'PredefinedExpr' is not supported
 18132 |             print_cell_data(block, indexNodeLeft, __func__, "left state");
       |                                                   ^~~~~~~~
test.orig.cpp:18133:121: warning: statement kind 'PredefinedExpr' is not supported
 18133 |             passSeriousError("/home/msagebaum/Kaiserslautern/mtu/Programms/trace_suite/trace/SOLVER/firstOrderSpace.c", __func__, 129, "ZONE %d at [%d %d %d] !!!\n", (block->zoneId), j, k, l); 

```

The NULL warning is probably hard to handle. What is you current point in that regard? From my view: If the primal code is null safe the derivative code should also be null safe. So NULL for the adjoint variable should not be a problem?

The warning about the modulo operator is kind of a false positive. The operation is just an index computation and does not enter the derivative computation. 

I looked for `PredefinedExpr` but it is not defined in my code. Is this something you add during the differentiation process?

There is still this fault, that I will try to get a reproducer tomorrow.
```
0.  Program arguments: /bin/clang++ -fno-crash-diagnostics -fplugin=/home/msagebaum/Applications/clad/inst/lib64/clad.so -I/home/msagebaum/Applications/clad/inst/include -Xclang -plugin-arg-clad -Xclang -fgenerate-source-file -DCLAD_NO_NUM_DIFF -I/home/msagebaum/Kaiserslautern/mtu/Programms/trace_suite/trace/INCLUDE -I/home/msagebaum/Kaiserslautern/mtu/Programms/trace_suite/build_jacobian_linear_dbg/trace -I/home/msagebaum/Applications/build/include -I/usr/include/openmpi-x86_64/ -I/home/msagebaum/Kaiserslautern/mtu/Programms/trace_suite/trace/AD -I/home/msagebaum/Kaiserslautern/mtu/Programms/trace_suite/contrib/ad_dependencies/adToolbox/include test.orig.cpp -c -o test.o -DNDEBUG -DSIMD_STRUCTS
1.  <eof> parser at end of file
 #0 0x00007f8d9562917a llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/lib64/libLLVM.so.20.1+0x22917a)
 #1 0x00007f8d95626104 llvm::sys::RunSignalHandlers() (/lib64/libLLVM.so.20.1+0x226104)
 #2 0x00007f8d9553f8f4 (/lib64/libLLVM.so.20.1+0x13f8f4)
 #3 0x00007f8d94e28070 __restore_rt (/lib64/libc.so.6+0x1a070)
 #4 0x00007f8d94e81e9c __pthread_kill_implementation (/lib64/libc.so.6+0x73e9c)
 #5 0x00007f8d94e27f3e gsignal (/lib64/libc.so.6+0x19f3e)
 #6 0x00007f8d94e0f6d0 abort (/lib64/libc.so.6+0x16d0)
 #7 0x00007f8d94e0f639 __assert_perror_fail (/lib64/libc.so.6+0x1639)
 #8 0x00007f8d942baae1 clang::FunctionDecl::getParamDecl(unsigned int) const /usr/include/clang/AST/Decl.h:0:5
 #9 0x00007f8d94518f14 clad::ReverseModeVisitor::VisitCallExpr(clang::CallExpr const*) /home/msagebaum/Applications/clad/lib/Differentiator/ReverseModeVisitor.cpp:1977:29
#10 0x00007f8d94500b18 clang::StmtVisitorBase<llvm::make_const_ptr, clad::ReverseModeVisitor, clad::StmtDiff>::Visit(clang::Stmt const*) /usr/include/clang/AST/StmtNodes.inc:614:1
#11 0x00007f8d944ff31e clad::ReverseModeVisitor::Visit(clang::Stmt const*, clang::Expr*) /home/msagebaum/Applications/clad/include/clad/Differentiator/ReverseModeVisitor.h:129:11
#12 0x00007f8d944fe69d clad::ReverseModeForwPassVisitor::ProcessSingleStmt(clang::Stmt const*) /home/msagebaum/Applications/clad/lib/Differentiator/ReverseModeForwPassVisitor.cpp:192:17
#13 0x00007f8d944fe761 clad::ReverseModeForwPassVisitor::VisitCompoundStmt(clang::CompoundStmt const*) /home/msagebaum/Applications/clad/lib/Differentiator/ReverseModeForwPassVisitor.cpp:201:29
#14 0x00007f8d94501898 clang::StmtVisitorBase<llvm::make_const_ptr, clad::ReverseModeVisitor, clad::StmtDiff>::Visit(clang::Stmt const*) /usr/include/clang/AST/StmtNodes.inc:1634:1
#15 0x00007f8d944ff31e clad::ReverseModeVisitor::Visit(clang::Stmt const*, clang::Expr*) /home/msagebaum/Applications/clad/include/clad/Differentiator/ReverseModeVisitor.h:129:11
#16 0x00007f8d9450fbb6 clad::ReverseModeVisitor::VisitIfStmt(clang::IfStmt const*)::$_0::operator()(clang::Stmt const*) const /home/msagebaum/Applications/clad/lib/Differentiator/ReverseModeVisitor.cpp:856:9
#17 0x00007f8d9450f471 clad::ReverseModeVisitor::VisitIfStmt(clang::IfStmt const*) /home/msagebaum/Applications/clad/lib/Differentiator/ReverseModeVisitor.cpp:875:37
#18 0x00007f8d945017e7 clang::StmtVisitorBase<llvm::make_const_ptr, clad::ReverseModeVisitor, clad::StmtDiff>::Visit(clang::Stmt const*) /usr/include/clang/AST/StmtNodes.inc:1586:1
#19 0x00007f8d944ff31e clad::ReverseModeVisitor::Visit(clang::Stmt const*, clang::Expr*) /home/msagebaum/Applications/clad/include/clad/Differentiator/ReverseModeVisitor.h:129:11
#20 0x00007f8d944fe69d clad::ReverseModeForwPassVisitor::ProcessSingleStmt(clang::Stmt const*) /home/msagebaum/Applications/clad/lib/Differentiator/ReverseModeForwPassVisitor.cpp:192:17
#21 0x00007f8d944fe761 clad::ReverseModeForwPassVisitor::VisitCompoundStmt(clang::CompoundStmt const*) /home/msagebaum/Applications/clad/lib/Differentiator/ReverseModeForwPassVisitor.cpp:201:29
#22 0x00007f8d94501898 clang::StmtVisitorBase<llvm::make_const_ptr, clad::ReverseModeVisitor, clad::StmtDiff>::Visit(clang::Stmt const*) /usr/include/clang/AST/StmtNodes.inc:1634:1
#23 0x00007f8d944ff31e clad::ReverseModeVisitor::Visit(clang::Stmt const*, clang::Expr*) /home/msagebaum/Applications/clad/include/clad/Differentiator/ReverseModeVisitor.h:129:11
#24 0x00007f8d9450fbb6 clad::ReverseModeVisitor::VisitIfStmt(clang::IfStmt const*)::$_0::operator()(clang::Stmt const*) const /home/msagebaum/Applications/clad/lib/Differentiator/ReverseModeVisitor.cpp:856:9
#25 0x00007f8d9450f471 clad::ReverseModeVisitor::VisitIfStmt(clang::IfStmt const*) /home/msagebaum/Applications/clad/lib/Differentiator/ReverseModeVisitor.cpp:875:37
#26 0x00007f8d945017e7 clang::StmtVisitorBase<llvm::make_const_ptr, clad::ReverseModeVisitor, clad::StmtDiff>::Visit(clang::Stmt const*) /usr/include/clang/AST/StmtNodes.inc:1586:1
#27 0x00007f8d944ff31e clad::ReverseModeVisitor::Visit(clang::Stmt const*, clang::Expr*) /home/msagebaum/Applications/clad/include/clad/Differentiator/ReverseModeVisitor.h:129:11
#28 0x00007f8d944fe69d clad::ReverseModeForwPassVisitor::ProcessSingleStmt(clang::Stmt const*) /home/msagebaum/Applications/clad/lib/Differentiator/ReverseModeForwPassVisitor.cpp:192:17
#29 0x00007f8d944fe761 clad::ReverseModeForwPassVisitor::VisitCompoundStmt(clang::CompoundStmt const*) /home/msagebaum/Applications/clad/lib/Differentiator/ReverseModeForwPassVisitor.cpp:201:29
#30 0x00007f8d94501898 clang::StmtVisitorBase<llvm::make_const_ptr, clad::ReverseModeVisitor, clad::StmtDiff>::Visit(clang::Stmt const*) /usr/include/clang/AST/StmtNodes.inc:1634:1
#31 0x00007f8d944ff31e clad::ReverseModeVisitor::Visit(clang::Stmt const*, clang::Expr*) /home/msagebaum/Applications/clad/include/clad/Differentiator/ReverseModeVisitor.h:129:11
#32 0x00007f8d944fd0ef clad::ReverseModeForwPassVisitor::Derive() /home/msagebaum/Applications/clad/lib/Differentiator/ReverseModeForwPassVisitor.cpp:70:25
#33 0x00007f8d9434ab9b clad::DerivativeBuilder::Derive(clad::DiffRequest const&) /home/msagebaum/Applications/clad/lib/Differentiator/DerivativeBuilder.cpp:568:18
#34 0x00007f8d9401da47 clad::plugin::CladPlugin::ProcessDiffRequest(clad::DiffRequest&) /home/msagebaum/Applications/clad/tools/ClangPlugin.cpp:323:52
#35 0x00007f8d9401ef6e clad::plugin::CladPlugin::FinalizeTranslationUnit() /home/msagebaum/Applications/clad/tools/ClangPlugin.cpp:536:11
#36 0x00007f8d9401fd44 clad::plugin::CladPlugin::HandleTranslationUnit(clang::ASTContext&) /home/msagebaum/Applications/clad/tools/ClangPlugin.cpp:577:9
#37 0x00007f8da055e2ec clang::MultiplexConsumer::HandleTranslationUnit(clang::ASTContext&) (/lib64/libclang-cpp.so.20.1+0x295e2ec)
#38 0x00007f8d9de12d49 clang::ParseAST(clang::Sema&, bool, bool) (/lib64/libclang-cpp.so.20.1+0x212d49)
#39 0x00007f8da05191af clang::FrontendAction::Execute() (/lib64/libclang-cpp.so.20.1+0x29191af)
#40 0x00007f8da047ba24 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/lib64/libclang-cpp.so.20.1+0x287ba24)
#41 0x00007f8da05accee clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/lib64/libclang-cpp.so.20.1+0x29accee)
#42 0x00005608bd367695 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/bin/clang+++0xa695)
#43 0x00005608bd363727 (/bin/clang+++0x6727)
#44 0x00007f8da001c0ad (/lib64/libclang-cpp.so.20.1+0x241c0ad)
#45 0x00007f8d9553f600 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/lib64/libLLVM.so.20.1+0x13f600)
#46 0x00007f8da001b965 clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (/lib64/libclang-cpp.so.20.1+0x241b965)
#47 0x00007f8d9ffd7bf9 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/lib64/libclang-cpp.so.20.1+0x23d7bf9)
#48 0x00007f8d9ffd7eb7 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/lib64/libclang-cpp.so.20.1+0x23d7eb7)
#49 0x00007f8d9fffaa09 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/lib64/libclang-cpp.so.20.1+0x23faa09)
#50 0x00005608bd362b54 clang_main(int, char**, llvm::ToolContext const&) (/bin/clang+++0x5b54)
#51 0x00005608bd3729a8 main (/bin/clang+++0x159a8)
#52 0x00007f8d94e11575 __libc_start_call_main (/lib64/libc.so.6+0x3575)
#53 0x00007f8d94e11628 __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x3628)
#54 0x00005608bd360e05 _start (/bin/clang+++0x3e05)

```

--- a/include/clad/Differentiator/BaseForwardModeVisitor.h
+++ b/include/clad/Differentiator/BaseForwardModeVisitor.h
@@ -124,6 +124,8 @@ public:
   StmtDiff VisitNullStmt(const clang::NullStmt* NS) { return StmtDiff{}; };
   StmtDiff
   VisitCXXStdInitializerListExpr(const clang::CXXStdInitializerListExpr* ILE);
+  StmtDiff VisitGNUNullExpr(const clang::GNUNullExpr* E);
+  StmtDiff VisitPredefinedExpr(const clang::PredefinedExpr* E);
 
   StmtDiff VisitOMPExecutableDirective(const clang::OMPExecutableDirective* D);
   StmtDiff VisitOMPParallelDirective(const clang::OMPParallelDirective* D);

--- a/include/clad/Differentiator/ReverseModeVisitor.h
+++ b/include/clad/Differentiator/ReverseModeVisitor.h
@@ -407,6 +407,8 @@ namespace clad {
     StmtDiff VisitForStmt(const clang::ForStmt* FS);
     StmtDiff VisitIfStmt(const clang::IfStmt* If);
     StmtDiff VisitImplicitCastExpr(const clang::ImplicitCastExpr* ICE);
+    StmtDiff VisitGNUNullExpr(const clang::GNUNullExpr* E);
+    StmtDiff VisitPredefinedExpr(const clang::PredefinedExpr* E);
 
 #if CLANG_VERSION_MAJOR > 16
     StmtDiff VisitLambdaExpr(const clang::LambdaExpr* LE);

--- a/lib/Differentiator/BaseForwardModeVisitor.cpp
+++ b/lib/Differentiator/BaseForwardModeVisitor.cpp
@@ -1612,6 +1612,21 @@ BaseForwardModeVisitor::VisitCStyleCastExpr(const CStyleCastExpr* CSCE) {
   return StmtDiff(castExpr, castExprDiff);
 }
 
+StmtDiff BaseForwardModeVisitor::VisitGNUNullExpr(const clang::GNUNullExpr* E) {
+  auto* Constant0 =
+      ConstantFolder::synthesizeLiteral(m_Context.IntTy, m_Context, /*val=*/0);
+  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+  return StmtDiff(const_cast<clang::GNUNullExpr*>(E), Constant0);
+}
+
+StmtDiff
+BaseForwardModeVisitor::VisitPredefinedExpr(const clang::PredefinedExpr* E) {
+  auto* Constant0 =
+      ConstantFolder::synthesizeLiteral(m_Context.IntTy, m_Context, /*val=*/0);
+  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+  return StmtDiff(const_cast<clang::PredefinedExpr*>(E), Constant0);
+}
+
 StmtDiff
 BaseForwardModeVisitor::VisitCXXNamedCastExpr(const CXXNamedCastExpr* NCE) {
   StmtDiff subExprDiff = Visit(NCE->getSubExpr());

--- a/lib/Differentiator/ReverseModeVisitor.cpp
+++ b/lib/Differentiator/ReverseModeVisitor.cpp
@@ -1931,10 +1931,6 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
       }
     }
 
-    // FIXME: Revisit this when variadic functions are supported.
-    if (FD->getNameAsString() == "printf" || FD->getNameAsString() == "fprintf")
-      return StmtDiff(Clone(CE));
-
     Expr* CUDAExecConfig = nullptr;
     if (const auto* KCE = dyn_cast<CUDAKernelCallExpr>(CE))
       CUDAExecConfig = Clone(KCE->getConfig());
@@ -2148,6 +2144,10 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
                      i = skip_this, e = CE->getNumArgs();
          i != e; ++i) {
       const Expr* arg = CE->getArg(i);
+      if ((i - skip_this) >= FD->getNumParams()) {
+        CallArgs.push_back(Clone(arg));
+        continue;
+      }
       const auto* PVD = FD->getParamDecl(i - skip_this);
       StmtDiff argDiff =
           DifferentiateCallArg(arg, PVD, PreCallStmts, /*isNonDiff=*/nonDiff,
@@ -2907,6 +2907,8 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
                         direction::reverse);
       auto* condDiffStored = IfStmtDiff.getRevSweepAsExpr();
       return BuildOp(BO_LAnd, condDiffStored, condVarRef);
+    } else if (opCode == BO_Rem) {
+      return BuildOp(opCode, Visit(L).getExpr(), Visit(R).getExpr());
     } else {
       // We should not output any warning on visiting boolean conditions
       // FIXME: We should support boolean differentiation or ignore it
@@ -3427,6 +3429,21 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
         ICE->getCastKind() == CK_LValueToRValue)
       result.updateStmtDx(getZeroInit(ICE->getType()));
     return result;
+  }
+
+  StmtDiff ReverseModeVisitor::VisitGNUNullExpr(const clang::GNUNullExpr* E) {
+    auto* Constant0 = ConstantFolder::synthesizeLiteral(m_Context.IntTy,
+                                                        m_Context, /*val=*/0);
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+    return StmtDiff(const_cast<clang::GNUNullExpr*>(E), Constant0);
+  }
+
+  StmtDiff
+  ReverseModeVisitor::VisitPredefinedExpr(const clang::PredefinedExpr* E) {
+    auto* Constant0 = ConstantFolder::synthesizeLiteral(m_Context.IntTy,
+                                                        m_Context, /*val=*/0);
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+    return StmtDiff(const_cast<clang::PredefinedExpr*>(E), Constant0);
   }
 
   StmtDiff ReverseModeVisitor::VisitCXXFunctionalCastExpr(

--- a/test/Regressions/issue-1682.cpp
+++ b/test/Regressions/issue-1682.cpp
@@ -1,0 +1,51 @@
+// RUN: %cladclang -fsyntax-only -Xclang -verify %s -I%S/../../include
+
+#include "clad/Differentiator/Differentiator.h"
+
+extern "C" int printf(const char*, ...);
+void a(...);
+int b; // expected-warning {{gradient uses a global variable 'b'; rerunning the gradient requires 'b' to be reset}}
+
+void c(float) {
+    // expected-warning@+2 {{attempted differentiation of function 'a' without definition and no suitable overload was found in namespace 'custom_derivatives'}}
+    // expected-note@+1 {{numerical differentiation is not viable for 'a'; considering 'a' as 0}}
+    a(b);
+}
+
+double f0(double x) {
+    int index = 10 % 3;
+    const char* func = __func__;
+    void* ptr = __null;
+
+    // expected-warning@+2 {{attempted differentiation of function 'printf' without definition and no suitable overload was found in namespace 'custom_derivatives'}}
+    // expected-note@+1 {{numerical differentiation is not viable for 'printf'; considering 'printf' as 0}}
+    printf("%f", x);
+
+    return x * x;
+}
+
+auto df = clad::gradient(f0);
+auto dg = clad::gradient(c);
+
+int b_fwd;
+
+void c_fwd(float) {
+    // expected-warning@+2 {{attempted differentiation of function 'a' without definition and no suitable overload was found in namespace 'custom_derivatives'}}
+    // expected-note@+1 {{numerical differentiation is not viable for 'a'; considering 'a' as 0}}
+    a(b_fwd);
+}
+
+double f0_fwd(double x) {
+    int index = 10 % 3;
+    const char* func = __func__;
+    void* ptr = __null;
+
+    // expected-warning@+2 {{attempted differentiation of function 'printf' without definition and no suitable overload was found in namespace 'custom_derivatives'}}
+    // expected-note@+1 {{numerical differentiation is not viable for 'printf'; considering 'printf' as 0}}
+    printf("%f", x);
+
+    return x * x;
+}
+
+auto df_fwd = clad::differentiate(f0_fwd);
+auto dg_fwd = clad::differentiate(c_fwd);

